src/theme.c

```diff
@@ -56,6 +57,43 @@ drop(struct lab_data_buffer **buffer)
 	}
 }
 
+static void
+create_hover_fallback(struct theme *theme, struct lab_data_buffer **hover_buffer,
+		struct lab_data_buffer *icon_buffer)
+{
+	assert(icon_buffer);
+	assert(!*hover_buffer);
+
+	struct surface_context icon =
+		get_cairo_surface_from_lab_data_buffer(icon_buffer);
+	int icon_width = cairo_image_surface_get_width(icon.surface);
+	int icon_height = cairo_image_surface_get_height(icon.surface);
+
+	/* FIXME: need to somehow respect rounded corners */
+	int width = MAX(SSD_BUTTON_WIDTH, icon_width);
+	int height = MAX(theme->title_height, icon_height);
```

Comment:
This may break the aspect ratio of the icon. Played around a bit with this and came up with this variant:
```diff
diff --git a/src/theme.c b/src/theme.c
index 6608b780..5a7a5913 100644
--- a/src/theme.c
+++ b/src/theme.c
@@ -69,9 +69,27 @@ create_hover_fallback(struct theme *theme, struct lab_data_buffer **hover_buffer
 	int icon_width = cairo_image_surface_get_width(icon.surface);
 	int icon_height = cairo_image_surface_get_height(icon.surface);
 
-	/* FIXME: need to somehow respect rounded corners */
-	int width = MAX(SSD_BUTTON_WIDTH, icon_width);
-	int height = MAX(theme->title_height, icon_height);
+	/* TODO: need to somehow respect rounded corners */
+	int width = SSD_BUTTON_WIDTH;
+	int height = theme->title_height;
+
+	if (width && height) {
+		/*
+		 * Maybe scale up the buffer proportionally.
+		 * It will get scaled down again by wlroots
+		 * when rendered and being required by the
+		 * current output scale.
+		 *
+		 * This ensures that icons > width or > height
+		 * keep their aspect ratio and are rendered the
+		 * same as without the hover overlay.
+		 */
+		double scale = MAX((double)icon_width / width, (double)icon_height / height);
+		if (scale > 1.0f) {
+			width = (double)width * scale;
+			height = (double)height * scale;
+		}
+	}
 
 	*hover_buffer = buffer_create_cairo(width, height, 1.0, true);
 ```

It is still not perfect:
- tried using `/usr/share/icons/hicolor/256x256/apps/chromium.png` as `max-active.png` and it still changing its aspect ratio slightly on hover
- the 6x6 default icons look even more blurry with output scale 2 when the hover is active, I played around with `CAIRO_ANTIALIAS_NONE` and `CAIRO_FILTER_FAST` but had no success.

Overall I'd consider this good enough to re-enable the fallback hover overlays.

Reply:
Thanks. Agree. Done (force-pushed against middle commit)